### ISSUE TITLE
Sixel support

### ISSUE THREAD

**Issue by atanunq** [atanunq/viuer#4]
[Sixel](https://en.wikipedia.org/wiki/Sixel) graphics are a great way to display images in high resolution, given that the terminal supports it. [libsixel](https://github.com/saitoha/libsixel) has Rust bindings that could be used to bring the functionality to `viuer`, see on [crates.io](https://crates.io/search?q=sixel).

The way to go about it would be to check for sixel compatibility in the main [print method](https://github.com/atanunq/viuer/blob/aeac70b5c99e9761161badec61ef1acaf5cd467c/src/lib.rs#L59-L69). If that is the case, call `SixelPrinter`, which lives in `sixel.rs` and implements the [`Printer` trait](https://github.com/atanunq/viuer/blob/aeac70b5c99e9761161badec61ef1acaf5cd467c/src/printer/mod.rs#L7-L9).

Just like with Kitty, rabite0's [hunter](https://github.com/rabite0/hunter) has a [sample implementation](https://github.com/rabite0/hunter/blob/355d9a3101f6d8dc375807de79e368602f1cb87d/src/hunter-media.rs#L714-L745) which could be extended.

**Comment by mmulet** [comment]
I saw the help wanted tag on https://github.com/atanunq/viu/issues/16 . I am looking for tasks to test out code relay, and this sounds simple enough. I will create a task on code relay to implement this, and we'll see how it goes.

**Comment by mmulet** [comment]
https://user-images.githubusercontent.com/6510497/103928877-0c0de400-50e2-11eb-9768-0097bc8a6ec4.mov

I've got sixel mostly working now. The only problem is with gif playback. 
It will draw the first frame in the correct position, but it will draw the rest of the frames in the top left corner of the window. 

Perhaps the problem lies in how the gif playback function is adjusting for scrollback.
viu app.rs: 236
```rust
 //keep replacing old pixels as the gif goes on so that scrollback
 // buffer is not filled (do not do that if it is the last frame of the gif
 // or a couple of files are being processed)
if iter.peek().is_some() || conf.loop_gif {
//since picture height is in pixel, we divide by 2 to get the height in
if let Err(crossterm::ErrorKind::IoError(e)) =
      execute!(stdout(), cursor::MoveUp(print_height as u16))
{
    if e.kind() == ErrorKind::BrokenPipe {
        //Stop printing. Output is probably piped to `head` or a similar tool
        break 'infinite;
    } else {
        return Err(e.into());
       }
 }
}

```
I've tried using libsixel's encode_from_file function, and that does work, however there are two major problems with that approach:
1. encode_from_file blocks the thread, and there is no way to check for ctrl-c. I tried running encode_from_file in another thread and polling for ctrlc in the main thread, and this does work 80% of the time. The other 20% percent of the time, it will freeze.
2. encode_from_file only works with some gifs. It will work fine with giphy.gif, but not all with trip.gif. It must be something o do with the gif decoder libsixel using internally.

Since, libsixel's encode_from_file does scroll back properly, I've taken a look at the code responsible. https://github.com/saitoha/libsixel/blob/master/src/tty.c . Look for sixel_tty_scroll.  I could port that over to rust, however it uses some unsafe, and unportable system calls like ioctl, so it would only work on unix systems.

Any thoughts? Encounter anything like this when using the other encoders?

**Comment by mmulet** [comment]
Ignore my last comment (I deleted it). I figured it out! I just have to test it on all the terminals that support sixel and then it's done.

**Comment by mmulet** [comment]
Works on xterm, mlterm and others in Linux.
Can't get the sixel-sys cargo package to compile on either windows or mac. Not sure if it's a sixel-sys issue or a libsixel issue, either way I'll open an issue somewhere and look into it. If you want we could merge sixel support for linux under a  cfg flag, or wait for windows and mac support.

**Comment by mmulet** [comment]
Found the issue, and created an workaround for sixel-sys: https://github.com/AdnoC/sixel-sys/issues/1 .
It hasn't been updated in 4 years, so not sure if the pull request will be accepted.
The bug is caused by a bug in the rust stdlib so it's possible that I could get it fixed in rust, so that we wouldn't have to pull in our own dependencies.

**Comment by atanunq** [comment]
Thank you so much for the effort! I don't have time to look at it right now but will aim to do so during the weekend.

**Comment by atanunq** [comment]
This is amazing! A lot of thought and effort has gone into it. Although, I am having troubles running it locally, I have a few comments:
- It is okay if you test only one or two of the most popular sixel-supporting emulators. The list you have laid out looks like an overkill to me
- [`adjust_offset`](https://github.com/atanunq/viuer/blob/80b15ec6d9d01c728ee3d6eca02a9be306a2a157/src/printer/mod.rs#L122-L147) and [`find_best_fit`](https://github.com/atanunq/viuer/blob/80b15ec6d9d01c728ee3d6eca02a9be306a2a157/src/printer/mod.rs#L60-L82) should be used to be able to 1. use the correct [x and y offsets](https://docs.rs/viuer/0.3.1/viuer/struct.Config.html#structfield.x), 2. choose the right image size based on Config's width and height. Correct me if I am wrong, but I don't see how these are handled.
- To add to the above, I think `get_pixel_size` can be removed and return the `(w, h)` from `find_best_fit`, see [here](https://github.com/atanunq/viuer/blob/80b15ec6d9d01c728ee3d6eca02a9be306a2a157/src/printer/kitty.rs#L123-L140).
- Checking for sixel support can be massively simplified. First, I'd rather use higher-level API for capturing the response, see [console::read_key](https://github.com/atanunq/viuer/blob/80b15ec6d9d01c728ee3d6eca02a9be306a2a157/src/printer/kitty.rs#L85-L96). For example, we might want to read everything until the character `c` is read and check whether it contains either `;4;` or `;4c`. This will simplify both reading and parsing and we won't have to debug term_info bitmaps.

As you can see, I really think we should keep the logic as simple as possible and as platform-agnostic as possible. This project should be able to stand on top of the lower level crates that handle the multi-platform complexities for us.

I really appreciate your work on this issue. Also, I am more than happy to work on the above mentioned points myself. Can you open a PR on which we can collaborate so that I can push code, too?

Thank you again

**Comment by mmulet** [comment]
Glad you like it! 
It's not just me, I used [coderelay.io](coderelay.io). Code relay is a platform for contributors to do a small part of an issue and then hand it off (like a relay race, you run a small leg and then pass the baton to the next runner). So, all the code relay users contributed to this (which at this point in time, is mainly me and my friends, but we are growing!). 

If you don't mind, it would really help us out if you put something like:

Made with coderelay.io 

or

Coderelay.io contributed to this

At the top of the README.md, so we can attract more contributors, so we can work on more open source projects.

Also, as an open source maintainer yourself, feel free to submit tasks to code relay:
[maintainer](https://www.coderelay.io/maintainer.html).

Fun fact: I made the code relay logo gif with viu,  I drew the running guy in blender then converted it to a gif, played it with viu on https://github.com/Swordfish90/cool-retro-term then captured it with obs-studio. This is why I wanted to help contribute to this project and give back, so to speak.

----

On to the actual issue:
1. Of those, Xterm is probably the most popular, which it did work on my PC. (When compiled to support it)
2. Yeah, looks like those aren't handled at all. I think it's just printing at image resolution right now
3. Probably, yeah.
4. Just looking for `;4;`, `;4c` should work fine. I was running into an issue with the Device Attributes query: Xterm would print the response to the console (It was visible to the user) instead of to stdout. So, that's the source of the platform specific code, just getting it to print to stdout. I looked, but didn't see a way to do that with crossterm. Maybe we should open an issue on crossterm itself and see if we can find a cross platform way to get the device attributes.

I created a pull-request.
